Lib/Inc/CPIDLoop.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file CPIDLoop.h
+ */
+
+/*
+ * CPIDLoop.h
+ *
+ *  Created on: 19 Feb 2022
+ *      Author: salavat.magazov
+ */
+
+#ifndef CPIDLOOP_H_
+#define CPIDLOOP_H_
+
+/* Default PID parameters.
+ * These are setup to make PID loop transparent. Output will be equal to error.
+ * Proportional is percentage of power per unit of error.
+ * Integral is units of output per unit of error per unit of time.
+ * Differential is unit of output per degree error change per unit of time.
+ * Normally for slow moving processes differential is zero.
+ */
+#define DEFAULT_P_COEFF 1
+#define DEFAULT_I_COEFF 0
+#define DEFAULT_D_COEFF 0
+
+class CPIDLoop
+{
+public:
+    struct CONTROL_STRUCT_T
+    {
+        CONTROL_STRUCT_T(float k_p, float k_i, float k_d);
+
+        float p_coeff;
+        float i_coeff;
+        float d_coeff;
+
+        float min_out;
+        float max_out;
+
+        float min_i;
+        float max_i;
+
+        float min_d;
+        float max_d;
+    };
+
+    CPIDLoop(float k_p = DEFAULT_P_COEFF,
+             float k_i = DEFAULT_I_COEFF,
+             float k_d = DEFAULT_D_COEFF);
+
+    float run(float target, float actual);
+    float run(float target, float actual, float deltaError);
```

Comment:
I don't get the point of having 2 run methods if the first one is calling the second one to get the output. Why not have something like the following:
```
float run(float target, float actual)
{
    float error = target - actual;
    float output  = 0;
    delta_error = error - m_old_error;
    m_old_error = error;
    ... continue with the second definition of run
    return output;
}
```
?

Reply:
This one is harder to explain. It is based on reality of the measuring input signals. Sometimes the signal you are trying to control is very noisy. D-term is very sensitive to noise because it, by the mathematical relationship associated with it, amplifies the changes and noise is very fast changes. If you want to filter the signal, then it makes sense for the filter to also calculate the rate of change of the signal and feed this filtered rate to the PID loop. 
I think it would probably make more sense to you if you knew differential calculus associated with control systems. It's not something you can learn from looking at the code or a quick google look up. If you are interested I'll recommend a math channel on youtube to watch. 
The code structure essentially allows you to either feed error value (which in this case is a representation of rate of change) or allow the algorithm to calculate it from the input signal based on past values. It also avoids repetition of code.